src/python/pants/backend/python/rules/pex.py

```diff
@@ -67,13 +74,81 @@ def create_from_adaptors(
         return PexInterpreterConstraints(constraint_set=tuple(sorted(interpreter_constraints)))
 
 
+@dataclass(frozen=True)
+class PexRequirementConstraints:
+    """A collection of paths to constraint files and/or Requirement-style strings."""
+
+    constraint_file_paths: Tuple[str, ...] = ()
+    raw_constraints: Tuple[str, ...] = ()
+    generated_file_name: ClassVar[str] = "requirement_constraints.generated.txt"
+
+    @classmethod
+    def create_from_raw_constraints(cls, constraints: Iterable[str]) -> "PexRequirementConstraints":
+        """The constraints may be paths to constraint files and/or requirement strings."""
+        constraint_file_paths = set()
+        raw_constraints = set()
+        # See https://www.python.org/dev/peps/pep-0440/#version-specifiers.
```

Comment:
What's the use case for supporting a list of literal constraints and not just files? In practice coming from outside pants, constraints live in dedicated checked in files, so its what folks will be used to. It's also harder than this, the contents of a constraints file need not have =, > or <:
```
$ cat constraints.txt 
git+https://github.com/pantsbuild/pex.git@v1.6.12#egg=pex

$ pip wheel --constraint constraints.txt pex
Collecting pex
  Cloning https://github.com/pantsbuild/pex.git (to revision v1.6.12) to /tmp/pip-wheel-k9khm3rq/pex
  Running command git clone -q https://github.com/pantsbuild/pex.git /tmp/pip-wheel-k9khm3rq/pex
  Running command git checkout -q 2ea34a9a2629e68bc636b14077e33d2987e99875
Building wheels for collected packages: pex
  Building wheel for pex (setup.py) ... done
  Created wheel for pex: filename=pex-1.6.12-py2.py3-none-any.whl size=765125 sha256=aa5a4615f197b57597abe8d1336cf6a84b967eac065a1caf10a9484248797284
  Stored in directory: /home/jsirois/dev/pantsbuild/jsirois-setup/tests
Successfully built pex

$ pex --constraint constraints.txt pex -cpex -opex-constraints.pex
$ ./pex-constraints.pex --version
pex-constraints.pex 1.6.12

$ pex --constraint constraints.txt pex==2.1.4 -cpex -opex-constraints.pex
$ ./pex-constraints.pex --version
pex-constraints.pex 1.6.12
```

Reply:
I think the usual use case will be that a company has a single global lockfile. In more complex cases they might have a small number of specific lockfiles. In even more complex cases they might want to be able to say "this target happens to add this extra constraint to the looser one in requirements.txt".  True, this is no longer a "lockfile" and more of a general "constraint", but that's OK.

In short, this seems analogous to interpreter constraints, and it makes sense to me to allow the same flexibility.  I'd rather not artificially constrain what can be done with this feature. 

Regarding detecting a constraint vs a file, that is a good point, since I guess a constraint can be, e.g., a VCS reference. So maybe we need some other, more principled way of distinguishing them. Maybe file references are prefixed with `@`? We already have that functionality for options, now that I think of it. So we could say that the option value (and the inline values in targets) are always lists of constraint strings, and if you want them to come from a file you use the existing @fromfile facility (which we would have to also apply to the inline constraints).

Does that make sense?

Reply:
I agree that there's value in making it lightweight to add a constraint "inline" rather than having to create a dedicated file. Per this discussion, I do think it's likely to have some targets need constraints that are different than the global constraints file: https://github.com/pantsbuild/pants/pull/9163#discussion_r383064563. 

Further, I can easily imagine only needing that more specific constraint for one specific target. In that case, it's much more ergonomic to in the BUILD file put `version==2.2` than to create a dedicated file with just that constraint.

I don't love the `@` syntax for referring to files, but it would solve the problem and is consistent with how we handle file references everywhere else.